### PR TITLE
[next] refactor/lifecycle hooks

### PR DESCRIPTION
**Pull request by BrunnerLivio**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[x] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?
At the moment the `nest-application-context.ts` is kind of bloated with lifecycle hooks. They can be easily refactored into dedicated files `packages/core/hooks/*.hook.ts`.

Related issue #1417 

## What is the new behavior?
The hooks are being refactored in a dedicated file per lifecycle hook (e.g. `on-module-init.hook.ts`).

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
I have not finished the PR yet. I want to get first some feedback before I continue. So please feel free to comment if this PR is in the alignment of the 6.0.0 roadmap.

Planned to refactor are for sure the lifecycle hooks:
- onModuleInit (done)
- onModuleDestroy
- onApplicationBootstrap

This is a preparation for #1417

### CODE REVIEW

File: packages/core/nest-application-context.ts [hunk 1]
```diff
@@ -76,44 +77,10 @@ export class NestApplicationContext implements INestApplicationContext {
...
-    const instances = [...module.controllers, ...providers];
-    const callOperator = (list: any) =>
-      list
-        .filter(instance => !isNil(instance))
-        .filter(this.hasOnModuleInitHook)
-        .map(async instance => (instance as OnModuleInit).onModuleInit());
-
-    await Promise.all(
-      callOperator(
-        iterate(instances)
```
Comment (BrunnerLivio) [inline packages/core/nest-application-context.ts]: Line 97 - 102 has been refactored to the function `getNonTransientInstances`, making it more consistent to `getTransientInstances` and better readable as seen in `on-module-init.hook.ts`

File: packages/core/injector/instance-trancient.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import iterate from 'iterare';
```
Comment (BrunnerLivio) [inline packages/core/injector/instance-trancient.ts]: Not quite sure if this is a good path to be in?

File: packages/core/injector/instance-trancient.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    .filter(item => !!item)
+    .map(({ instance }: any) => instance)
+    .toArray() as InstanceWrapper[];
+}
+
+/**
+ * Returns the instances which are not transient
+ * @param instances The instances which should be checked whether they are transcient
+ */
+export function getNonTransientInstances(
```
Comment (BrunnerLivio) [inline packages/core/injector/instance-trancient.ts]: Not quite sure on that naming...

File: packages/core/hooks/on-module-init.hook.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import iterate from 'iterare';
+
+import { OnModuleInit } from '@nestjs/common';
+import { isNil, isUndefined } from '@nestjs/common/utils/shared.utils';
+
+import { Module } from '../injector/module';
+import { getTransientInstances, getNonTransientInstances } from '../injector/instance-trancient';
+import { InstanceWrapper } from 'injector/instance-wrapper';
```
Comment (kamilmysliwiec) [inline packages/core/hooks/on-module-init.hook.ts]: This absolute import will be broken after compilation

File: packages/core/injector/instance-trancient.ts (hunk 3 above)
Comment (kamilmysliwiec) [reply packages/core/injector/instance-trancient.ts]: Sounds fair enough

**Comment by kamilmysliwiec** [comment]
> At the moment the nest-application-context.ts is kind of bloated with lifecycle hooks. They can be easily refactored into dedicated files packages/core/hooks/*.hook.ts.

I like the idea!

**Comment by BrunnerLivio** [comment]
Finished from my side :)

**Comment by shekohex** [comment]
I know that is not the place to discuss this, but can't we use Events for that kind of hooks ? Something like nodejs event emitter or something more like eventemitter2 !

One use case, i wanna leverage nest in functional way where i use functions for handle requests .. etc without needing for classes .

I think I'll open a new issue to discuss that.

**Comment by BrunnerLivio** [comment]
Added the onApplicationShutdown hook #1417

**Comment by marcus-sa** [comment]
@BrunnerLivio it is the expected behavios. You need to manually call the hooks yourself whenever fit appropriate in the test/e2e cycle.

**Comment by BrunnerLivio** [comment]
@marcus-sa oh yes, of course... Thanks ^^

File: sample/01-cats-app/src/cats/cats.service.ts [hunk 5]
```diff
@@ -1,8 +1,17 @@
...
-export class CatsService {
+export class CatsService implements OnApplicationShutdown, OnModuleDestroy {
+  onApplicationShutdown(signal: string) {
+    console.log(signal);
+  }
+
+
+  onModuleDestroy() {
+    console.log('onModuleDestroy');
+  }
```
Comment (kamilmysliwiec) [inline sample/01-cats-app/src/cats/cats.service.ts]: It looks like a side-effect :D

Comment (BrunnerLivio) [reply sample/01-cats-app/src/cats/cats.service.ts]: Fixed :)

**Comment by BrunnerLivio** [comment]
@kamilmysliwiec just realized that the folder `src/core/hooks/` for the lifecycle hooks could be confusing if #1407 gets merged, because the async hooks uses the folder `src/common/hooks`. Should I rename the folder to `src/core/lifecycle-hooks`?

**Comment by kamilmysliwiec** [comment]
LGTM

**Pull request merged**